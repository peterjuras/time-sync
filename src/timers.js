import { SECONDS } from "./constants";
import { generateId } from "./lib/id";
import { validateInterval } from "./lib/validation";
import { getMs } from "./lib/get-ms";

const DEFAULT_TIMER_CONFIG = {
  interval: SECONDS,
  unit: 1
};

function validateTimerConfig(timerConfig) {
  validateInterval(timerConfig);

  if (
    typeof timerConfig.unit !== "undefined" &&
    (!Number.isInteger(timerConfig.unit) || timerConfig.unit <= 0)
  ) {
    throw new Error("unit must be a positive integer");
  }
}

function validateAddTimerArgs(callback, timerConfig) {
  if (!callback) {
    throw new Error(
      "You need to provide a callback as the first argument to addTimer"
    );
  }

  validateTimerConfig(timerConfig);
}

function roundTick(ms, tick) {
  return tick - (tick % ms);
}

function getCurrentTick(ms) {
  return roundTick(ms, Date.now());
}

function getNextTick({ ms }, time) {
  const newTime = time + ms;
  return roundTick(ms, newTime);
}

function getUnixTimeStamp(tick) {
  return Math.floor(tick / 1000);
}

export function getCurrentTime(timerConfig = {}) {
  validateTimerConfig(timerConfig);

  const config = {
    ...DEFAULT_TIMER_CONFIG,
    ...timerConfig
  };

  return getUnixTimeStamp(getCurrentTick(getMs(config)));
}

export class Timers {
  timers = {};

  currentTimeout;

  nextTick;

  revalidate = () => {
    clearTimeout(this.currentTimeout);
    const now = Date.now();

    this.nextTick = Object.keys(this.timers).reduce((prev, timerId) => {
      const timer = this.timers[timerId];

      if (timer.nextTick <= now) {
        const usedTime = roundTick(timer.ms, now);
        timer.callback(getUnixTimeStamp(usedTime));
        timer.nextTick = getNextTick(timer, now);
      }

      if (!prev || prev > timer.nextTick) {
        return timer.nextTick;
      }
      return prev;
    }, null);

    if (this.nextTick) {
      const nextTickDelta = this.nextTick - Date.now();
      this.currentTimeout = setTimeout(this.revalidate, nextTickDelta);
    }
  };

  removeTimer = id => {
    const timer = this.timers[id];
    if (!timer) {
      return;
    }

    delete this.timers[id];
    if (timer.nextTick === this.nextTick) {
      this.revalidate();
    }
  };

  removeAllTimers = () => {
    this.timers = {};
    this.revalidate();
  };

  addTimer = (callback, timerConfig = {}) => {
    validateAddTimerArgs(callback, timerConfig);

    const id = generateId();
    const newTimer = {
      ...DEFAULT_TIMER_CONFIG,
      ...timerConfig,
      callback,
      id
    };
    newTimer.ms = getMs(newTimer);
    newTimer.nextTick = getNextTick(newTimer, Date.now());

    this.timers[id] = newTimer;

    if (!this.nextTick || newTimer.nextTick < this.nextTick) {
      this.revalidate();
    }

    return () => this.removeTimer(id);
  };
}
