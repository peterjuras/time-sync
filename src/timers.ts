import { Interval } from "./constants";
import { generateId } from "./lib/id";
import { validateInterval } from "./lib/validation";
import { getMs } from "./lib/get-ms";

interface TimerConfig {
  interval: Interval;
  unit: number;
}

const DEFAULT_TIMER_CONFIG: TimerConfig = {
  interval: Interval.SECONDS,
  unit: 1
};

type TimeCallback = (currentTime: number) => void;

interface StoredTimerConfig extends TimerConfig {
  id: number;
  callback: TimeCallback;
  ms: number;
  nextTick: number;
}

function validateTimerConfig(timerConfig: Partial<TimerConfig>): void {
  validateInterval(timerConfig.interval);

  if (
    typeof timerConfig.unit !== "undefined" &&
    (!Number.isInteger(timerConfig.unit) || timerConfig.unit <= 0)
  ) {
    throw new Error("unit must be a positive integer");
  }
}

function validateAddTimerArgs(
  callback: TimeCallback,
  timerConfig: Partial<TimerConfig>
): void {
  if (!callback) {
    throw new Error(
      "You need to provide a callback as the first argument to addTimer"
    );
  }

  validateTimerConfig(timerConfig);
}

function getTimezoneOffset(): number {
  return new Date().getTimezoneOffset() * 60 * 1000;
}

function roundTick(ms: number, tick: number, interval?: Interval): number {
  let delta = tick % ms;
  if (interval === Interval.DAYS) {
    const timezoneDelta = delta - getTimezoneOffset();
    if (timezoneDelta > ms) {
      return tick + timezoneDelta;
    }
    delta = Math.abs(timezoneDelta);
  }

  return tick - delta;
}

function getCurrentTick(ms: number, interval: Interval): number {
  return roundTick(ms, Date.now(), interval);
}

function getNextTick(interval: Interval, ms: number, time: number): number {
  const newTime = time + ms;
  return roundTick(ms, newTime, interval);
}

function getUnixTimeStamp(tick: number): number {
  return Math.floor(tick / 1000);
}

export function getCurrentTime(timerConfig: Partial<TimerConfig> = {}): number {
  validateTimerConfig(timerConfig);

  const config: TimerConfig = {
    ...DEFAULT_TIMER_CONFIG,
    ...timerConfig
  };

  return getUnixTimeStamp(
    getCurrentTick(getMs(config.interval, config.unit), config.interval)
  );
}

export class Timers {
  private timers: {
    [key: string]: StoredTimerConfig;
  } = {};

  private currentTimeout?: number;

  private nextTick = 0;

  public removeAllTimers = (): void => {
    this.timers = {};
    this.revalidate();
  };

  public addTimer = (
    callback: TimeCallback,
    timerConfig: Partial<TimerConfig> = {}
  ): (() => void) => {
    validateAddTimerArgs(callback, timerConfig);

    const id = generateId();
    const newTimerBase = {
      ...DEFAULT_TIMER_CONFIG,
      ...timerConfig
    };
    const ms = getMs(newTimerBase.interval, newTimerBase.unit);
    const newTimer = {
      ...newTimerBase,
      callback,
      id,
      ms,
      nextTick: getNextTick(newTimerBase.interval, ms, Date.now())
    };

    this.timers[id] = newTimer;

    if (!this.nextTick || newTimer.nextTick < this.nextTick) {
      this.revalidate();
    }

    return (): void => this.removeTimer(id);
  };

  public revalidate = (): void => {
    clearTimeout(this.currentTimeout);
    const now = Date.now();

    this.nextTick = Object.keys(this.timers).reduce(
      (prev: number, timerId: string): number => {
        const timer = this.timers[timerId];
        if (!timer) {
          // Timer was already removed within callback of another timer
          return prev;
        }

        if (timer.nextTick <= now) {
          const usedTime = roundTick(timer.ms, now, timer.interval);
          timer.callback(getUnixTimeStamp(usedTime));
          timer.nextTick = getNextTick(timer.interval, timer.ms, now);
        }

        if (!prev || prev > timer.nextTick) {
          return timer.nextTick;
        }
        return prev;
      },
      0
    );

    if (this.nextTick) {
      const nextTickDelta = this.nextTick - Date.now();
      this.currentTimeout = setTimeout(this.revalidate, nextTickDelta);
    }
  };

  private removeTimer = (id: number): void => {
    const timer = this.timers[id];
    if (!timer) {
      return;
    }

    delete this.timers[id];
    if (timer.nextTick === this.nextTick) {
      this.revalidate();
    }
  };
}
