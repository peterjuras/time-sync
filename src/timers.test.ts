import TimeSync from ".";
import { Timers } from "./timers";
import FakeTimers from "@sinonjs/fake-timers";

describe("#timers", () => {
  const instance = new Timers();
  let clock: FakeTimers.InstalledClock;

  beforeEach(() => {
    clock = FakeTimers.install({
      now: 10001,
    });
  });

  afterEach(() => {
    clock.uninstall();
    instance.removeAllTimers();
  });

  describe("#addTimer", () => {
    it("should be exported correctly", () => {
      expect(instance.addTimer).toBeInstanceOf(Function);
    });

    it("should throw if no callback is provided", () => {
      expect(instance.addTimer).toThrowErrorMatchingSnapshot();
    });

    it("should throw if for an unknown interval", () => {
      expect(() =>
        instance.addTimer(jest.fn(), {
          interval: [] as any,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.addTimer(jest.fn(), {
          interval: {} as any,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.addTimer(jest.fn(), {
          interval: 123 as any,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.addTimer(jest.fn(), {
          interval: "" as any,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.addTimer(jest.fn(), {
          interval: "a" as any,
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it("should throw if for an invalid unit", () => {
      expect(() =>
        instance.addTimer(jest.fn(), {
          unit: [] as any,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.addTimer(jest.fn(), {
          unit: {} as any,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.addTimer(jest.fn(), {
          unit: "" as any,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.addTimer(jest.fn(), {
          unit: "a" as any,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.addTimer(jest.fn(), {
          unit: "15" as any,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.addTimer(jest.fn(), {
          unit: 250.291,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.addTimer(jest.fn(), {
          unit: 0,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.addTimer(jest.fn(), {
          unit: -12,
        })
      ).toThrowErrorMatchingSnapshot();

      expect(() =>
        instance.addTimer(jest.fn(), {
          unit: -12.28,
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it("should add a second timer by default", () => {
      const mock = jest.fn();
      instance.addTimer(mock);

      expect(mock).toHaveBeenCalledTimes(0);
      clock.tick(998);
      expect(mock).toHaveBeenCalledTimes(0);

      clock.tick(1);
      expect(mock).toHaveBeenCalledTimes(1);

      clock.tick(999);
      expect(mock).toHaveBeenCalledTimes(1);
    });

    it("should correctly callback for a second timer", () => {
      const mock = jest.fn();
      instance.addTimer(mock, { interval: TimeSync.SECONDS });

      expect(mock).toHaveBeenCalledTimes(0);
      clock.tick(1000 * 2 - 2);

      expect(mock).toHaveBeenCalledTimes(1);
      expect(mock).toHaveBeenLastCalledWith(11);
    });

    it("should correctly callback for a minute timer", () => {
      const mock = jest.fn();
      instance.addTimer(mock, { interval: TimeSync.MINUTES });

      expect(mock).toHaveBeenCalledTimes(0);
      clock.tick(60000 * 2 - 2);

      expect(mock).toHaveBeenCalledTimes(2);
      expect(mock).toHaveBeenLastCalledWith(120);
    });

    it("should correctly callback for a hour timer", () => {
      const mock = jest.fn();
      instance.addTimer(mock, { interval: TimeSync.HOURS });

      expect(mock).toHaveBeenCalledTimes(0);
      clock.tick(1000 * 60 * 60 * 2 - 2);

      expect(mock).toHaveBeenCalledTimes(2);
      expect(mock).toHaveBeenLastCalledWith(60 * 60 * 2);
    });

    it("should correctly callback for a day timer", () => {
      const mock = jest.fn();
      instance.addTimer(mock, { interval: TimeSync.DAYS });

      expect(mock).toHaveBeenCalledTimes(0);
      clock.tick(
        1000 * 60 * 60 * 24 * 2 - 2 - new Date().getTimezoneOffset() * 60 * 1000
      );

      expect(mock).toHaveBeenCalledTimes(2);
      expect(mock).toHaveBeenLastCalledWith(
        60 * 60 * 24 * 2 + new Date().getTimezoneOffset() * 60
      );
    });

    it("should work with a mix of timers", () => {
      const mockSecond = jest.fn();
      const mockTenSeconds = jest.fn();
      const mockMinute = jest.fn();
      instance.addTimer(mockSecond);
      instance.addTimer(mockTenSeconds, { unit: 10 });
      instance.addTimer(mockMinute, { interval: TimeSync.MINUTES });

      clock.tick(1000 * 60 - 1);

      expect(mockSecond).toHaveBeenCalledTimes(60);
      expect(mockSecond.mock.calls).toMatchSnapshot();

      expect(mockTenSeconds).toHaveBeenCalledTimes(6);
      expect(mockTenSeconds.mock.calls).toMatchSnapshot();

      expect(mockMinute).toHaveBeenCalledTimes(1);
      expect(mockMinute.mock.calls).toMatchSnapshot();
    });

    it("should clean up timeouts if the last timer is removed", () => {
      const mock = jest.fn();
      const ref = instance.addTimer(mock);

      clock.tick(1000);
      expect(mock).toHaveBeenCalledTimes(1);
      ref();

      clock.tick(1000);
      expect(mock).toHaveBeenCalledTimes(1);
    });

    it("should correctly remove timer if it schedules for the next tick", () => {
      const mock1 = jest.fn();
      const mock2 = jest.fn();
      const mock3 = jest.fn();
      const mock4 = jest.fn();
      instance.addTimer(mock1);
      const removeReference = instance.addTimer(mock2);
      instance.addTimer(mock3, { interval: TimeSync.MINUTES });
      instance.addTimer(mock4);

      // Advance to just before two minute
      clock.tick(1000 * 60 - 2 - 10000);

      expect(mock1).toHaveBeenCalledTimes(49);
      expect(mock2).toHaveBeenCalledTimes(49);
      expect(mock3).toHaveBeenCalledTimes(0);
      expect(mock4).toHaveBeenCalledTimes(49);

      removeReference();

      clock.tick(200);

      expect(mock1).toHaveBeenCalledTimes(50);
      expect(mock2).toHaveBeenCalledTimes(49);
      expect(mock3).toHaveBeenCalledTimes(1);
      expect(mock4).toHaveBeenCalledTimes(50);
    });

    it("removing a timer should not throw if it is called after all timers have been removed", () => {
      const mock = jest.fn();
      const ref = instance.addTimer(mock);

      instance.removeAllTimers();
      ref();
    });

    it("removing a timer should not throw if it is called multiple times", () => {
      const mock = jest.fn();
      const ref = instance.addTimer(mock);

      ref();
      ref();
      ref();
    });
  });

  describe("#getCurrentTime", () => {
    it("should be exported correctly", () => {
      expect(TimeSync.getCurrentTime).toBeInstanceOf(Function);
    });

    it("should give the correctly rounded value", () => {
      expect(TimeSync.getCurrentTime()).toBe(10);
      clock.tick(1250);
      expect(TimeSync.getCurrentTime()).toBe(11);
    });

    it("should work for defined timer configurations", () => {
      const timerConfig = { unit: 90, interval: TimeSync.MINUTES };
      expect(TimeSync.getCurrentTime(timerConfig)).toBe(0);
      clock.tick(1000 * 60 * 60 * 2);
      expect(TimeSync.getCurrentTime(timerConfig)).toBe(60 * 90);
    });

    it("should return adjusted timezone value for DAYS configurations", () => {
      const timerConfig = { interval: TimeSync.DAYS };
      const timezoneOffset = new Date().getTimezoneOffset() * 60;
      expect(TimeSync.getCurrentTime(timerConfig)).toBe(timezoneOffset);
      clock.tick(1000 * 60 * 60 * 24 * 2);
      expect(TimeSync.getCurrentTime(timerConfig)).toBe(
        24 * 60 * 60 * 2 + timezoneOffset
      );
    });
  });

  describe("#removeAllTimers", () => {
    it("should be exported correctly", () => {
      expect(instance.removeAllTimers).toBeInstanceOf(Function);
    });
  });

  describe("#revalidateAllTimers", () => {
    it("should be exported correctly", () => {
      expect(instance.revalidateAllTimers).toBeInstanceOf(Function);
    });

    it("should fire all necessary timers if system clock has changed", () => {
      const mockSecond = jest.fn();
      const mockTenSeconds = jest.fn();
      const mockMinute = jest.fn();
      instance.addTimer(mockSecond);
      instance.addTimer(mockTenSeconds, { unit: 10 });
      instance.addTimer(mockMinute, { interval: TimeSync.MINUTES });

      clock.setSystemTime(1000 * 60 * 60);

      expect(mockSecond).toHaveBeenCalledTimes(0);
      expect(mockTenSeconds).toHaveBeenCalledTimes(0);
      expect(mockMinute).toHaveBeenCalledTimes(0);

      clock.tick(1);

      expect(mockSecond).toHaveBeenCalledTimes(0);
      expect(mockTenSeconds).toHaveBeenCalledTimes(0);
      expect(mockMinute).toHaveBeenCalledTimes(0);

      instance.revalidateAllTimers();
      clock.tick(1);

      expect(mockSecond).toHaveBeenCalledTimes(1);
      expect(mockTenSeconds).toHaveBeenCalledTimes(1);
      expect(mockMinute).toHaveBeenCalledTimes(1);
    });

    it("should revalidate even if next timer has not fired yet", () => {
      const mock = jest.fn();
      instance.addTimer(mock);

      clock.setSystemTime(10250);
      instance.revalidateAllTimers();

      clock.tick(750);
      expect(mock).toHaveBeenCalledTimes(1);
      expect(mock).toHaveBeenLastCalledWith(11);
    });

    it("should resync automatically after any timer is fired", () => {
      const mock1 = jest.fn();
      const mock2 = jest.fn();
      const mock3 = jest.fn();
      instance.addTimer(mock1);
      instance.addTimer(mock2, { unit: 2 });
      instance.addTimer(mock3, { unit: 2 });

      clock.setSystemTime(12500);
      clock.tick(999);

      expect(mock1).toHaveBeenCalledTimes(1);
      expect(mock2).toHaveBeenCalledTimes(1);
      expect(mock3).toHaveBeenCalledTimes(1);
    });

    it("should not automatically revalidate after a future timer has been removed", () => {
      const mock1 = jest.fn();
      const mock2 = jest.fn();

      const ref = instance.addTimer(mock1, { interval: TimeSync.MINUTES });
      instance.addTimer(mock2);

      clock.setSystemTime(1250);
      ref();

      expect(mock2).toHaveBeenCalledTimes(0);
    });

    it("should not break if removal of timer removes second timer", () => {
      // eslint-disable-next-line prefer-const
      let removeSecondTimerRef: () => void;

      instance.addTimer(() => {
        removeSecondTimerRef();
      });
      removeSecondTimerRef = instance.addTimer(jest.fn());

      clock.tick(1000);
    });
  });
});
