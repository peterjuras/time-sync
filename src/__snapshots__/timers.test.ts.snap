// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#timers #addTimer should throw if for an invalid unit 1`] = `"unit must be a positive integer"`;

exports[`#timers #addTimer should throw if for an invalid unit 2`] = `"unit must be a positive integer"`;

exports[`#timers #addTimer should throw if for an invalid unit 3`] = `"unit must be a positive integer"`;

exports[`#timers #addTimer should throw if for an invalid unit 4`] = `"unit must be a positive integer"`;

exports[`#timers #addTimer should throw if for an invalid unit 5`] = `"unit must be a positive integer"`;

exports[`#timers #addTimer should throw if for an invalid unit 6`] = `"unit must be a positive integer"`;

exports[`#timers #addTimer should throw if for an invalid unit 7`] = `"unit must be a positive integer"`;

exports[`#timers #addTimer should throw if for an invalid unit 8`] = `"unit must be a positive integer"`;

exports[`#timers #addTimer should throw if for an invalid unit 9`] = `"unit must be a positive integer"`;

exports[`#timers #addTimer should throw if for an unknown interval 1`] = `"interval must be one of TimeSync.SECONDS, TimeSync.MINUTES, TimeSync.HOURS or TimeSync.DAYS"`;

exports[`#timers #addTimer should throw if for an unknown interval 2`] = `"interval must be one of TimeSync.SECONDS, TimeSync.MINUTES, TimeSync.HOURS or TimeSync.DAYS"`;

exports[`#timers #addTimer should throw if for an unknown interval 3`] = `"interval must be one of TimeSync.SECONDS, TimeSync.MINUTES, TimeSync.HOURS or TimeSync.DAYS"`;

exports[`#timers #addTimer should throw if for an unknown interval 4`] = `"interval must be one of TimeSync.SECONDS, TimeSync.MINUTES, TimeSync.HOURS or TimeSync.DAYS"`;

exports[`#timers #addTimer should throw if for an unknown interval 5`] = `"interval must be one of TimeSync.SECONDS, TimeSync.MINUTES, TimeSync.HOURS or TimeSync.DAYS"`;

exports[`#timers #addTimer should throw if no callback is provided 1`] = `"You need to provide a callback as the first argument to addTimer"`;

exports[`#timers #addTimer should work with a mix of timers 1`] = `
[
  [
    11,
  ],
  [
    12,
  ],
  [
    13,
  ],
  [
    14,
  ],
  [
    15,
  ],
  [
    16,
  ],
  [
    17,
  ],
  [
    18,
  ],
  [
    19,
  ],
  [
    20,
  ],
  [
    21,
  ],
  [
    22,
  ],
  [
    23,
  ],
  [
    24,
  ],
  [
    25,
  ],
  [
    26,
  ],
  [
    27,
  ],
  [
    28,
  ],
  [
    29,
  ],
  [
    30,
  ],
  [
    31,
  ],
  [
    32,
  ],
  [
    33,
  ],
  [
    34,
  ],
  [
    35,
  ],
  [
    36,
  ],
  [
    37,
  ],
  [
    38,
  ],
  [
    39,
  ],
  [
    40,
  ],
  [
    41,
  ],
  [
    42,
  ],
  [
    43,
  ],
  [
    44,
  ],
  [
    45,
  ],
  [
    46,
  ],
  [
    47,
  ],
  [
    48,
  ],
  [
    49,
  ],
  [
    50,
  ],
  [
    51,
  ],
  [
    52,
  ],
  [
    53,
  ],
  [
    54,
  ],
  [
    55,
  ],
  [
    56,
  ],
  [
    57,
  ],
  [
    58,
  ],
  [
    59,
  ],
  [
    60,
  ],
  [
    61,
  ],
  [
    62,
  ],
  [
    63,
  ],
  [
    64,
  ],
  [
    65,
  ],
  [
    66,
  ],
  [
    67,
  ],
  [
    68,
  ],
  [
    69,
  ],
  [
    70,
  ],
]
`;

exports[`#timers #addTimer should work with a mix of timers 2`] = `
[
  [
    20,
  ],
  [
    30,
  ],
  [
    40,
  ],
  [
    50,
  ],
  [
    60,
  ],
  [
    70,
  ],
]
`;

exports[`#timers #addTimer should work with a mix of timers 3`] = `
[
  [
    60,
  ],
]
`;
